src/extension-support/extension-manager.js

```diff
@@ -242,8 +242,34 @@ class ExtensionManager {
             }
             return result;
         }, []);
+        extensionInfo.menus = extensionInfo.menus || [];
+        extensionInfo.menus = this._prepareMenuInfo(serviceName, extensionInfo.menus);
         return extensionInfo;
     }
+    
+    /**
+     * Prepare extension menus. e.g. setup binding for dynamic menu functions.
+     * @param {string} serviceName - the name of the service hosting this extension block
+     * @param {Array.<MenuInfo>} menus - the menu defined by the extension.
+     * @returns {Array.<MenuInfo>} - a menuInfo object with all preprocessing done.
+     * @private
+     */
+    _prepareMenuInfo (serviceName, menus) {
+        const menuNames = Object.getOwnPropertyNames(menus);
+        for (let i = 0; i < menuNames.length; i++) {
+            const item = menuNames[i];
+            // If the value is a string, it should be the name of a function in the
+            // extension object to call to populate the menu whenever it is opened.
+            // Set up the binding for the function object here so
+            // we can use it later when converting the menu for Scratch Blocks.
+            if (typeof menus[item] === 'string') {
+                const serviceObject = dispatch.services[serviceName];
+                const menuFunc = serviceObject[menus[item]].bind(serviceObject);
```

Comment:
Ideally this should be something like `dispatch.call.bind(dispatch, serviceName, menus[item])` but that's only OK if you can consume a promise for the menu items. If that's not easy then I think this is fine for now. We'll need to figure out how to deal with worker extensions later, though.

Reply:
I dug into this a little and the problem I'm having right now is that it actually errors out in the blocks code while trying to replace the menu (whether I return a promise from the extension's menu build function or not).  I tried returning a few different formats for the menu but haven't had any luck yet.  It is tricky to debug since the blocks code is compressed into a giant blob. Is there a way to run the gui with uncompressed blocks?  That would make it much easier to tell if I'm on the right track or in a rabbit hole :)

Reply:
As discussed, punting on this for now since scratch-blocks doesn't currently accept the promise or have a mechanism to update the menu outside of the function called when the menu is clicked. Added a TODO.